File: ps-2019-assignment-03-Bessem11-master/task-1/master-mild.js

var extractsecretcode= function(secretCodeLocation,func) {
    try{
    var fs = require('fs');
    var inputcode = fs.readFileSync(secretCodeLocation).toString();
    var secretcode= inputcode.trim();
    var lengthdiff = 0;
    }catch(err){
    console.log('No such file name or directory');
    process.exit()
    }
    var digits=4

    if(secretcode.length<digits)
    { 
        lengthdiff=digits-secretcode.length

        for (i=0;i<lengthdiff;i++)
        {
            secretcode='0'+secretcode;
        }

    }
    else if (secretcode.length>digits)
    {
        lengthdiff=secretcode.length-digits;
        secretcode=secretcode.substr(lengthdiff,secretcode.length);
    }
    // inputGuessedCode
    func(secretcode,verifyinput)
}

// ALESSIO: Name of this function is misleading? It seems that you are accumulating all the guessedCodes into
//      codes to call the next function with it
var inputGuessedCode=async function(secretcode,func)
{
    const readline = require('readline');
    codes=[];

    this.cons = readline.createInterface({
        input: process.stdin,
        prompt: ''
    });

    this.codesc = [];

    this.promise = new Promise(res => {
        this.resolve = res;
    });

    this.cons.on('line', (line) => {
        this.resolve();
        this.codesc.push(line);
        this.promise = new Promise(res => {
        this.resolve = res;
        });
    });

    if (this.codesc.length == 0) {
        await this.promise;
    }    

    codes.push(codesc);
    codes.push(secretcode);
    // the codes variable contain the guessedcodes and the secret code
    //it's format is [[guessedcode1,guessedcode2,....],secretcode]
    //here at the end of this function we call the next function

    //verifyinput
    func(codes,compute);

}

var verifyinput= async function(codesc,func){

    // ALESSIO Why do you need await here?
    var codes=await codesc
    var guessedcodes=codes[0];
    var digits=4

    for(var i=0;i<guessedcodes.length;i++)
    {
        var nospacescode = guessedcodes[i].trim(); 
        var isnumber=true;  
        
        for ( var m=0;m<nospacescode.length&&isnumber;m++)
           {
                var c= nospacescode.charAt(m);
                if (!(c >= '0' && c <= '9')) //verifying if the user input consists only of digits
                {
                     isnumber=false;
                } 
            }

        if (nospacescode.length!=digits || !isnumber)
            {
                //if the guessedcode is invalid then we replace the guessed code with [guessedcode1,null]
                guessedcodes[i]=[guessedcodes[i],null]
            }
    }
    codes[0]=guessedcodes;

    //if for exemple the guessedcode1 is invalid , codes will have this format [[[guessedcode1,null],guesssedcode2,..],secretcode]
     //here at the end of this function we call the next function

     // ALESSIO Why do not you simply generate the output as you go instead of accumulating them ?
     // compute
    func(codes,print)
}

var compute=async function(codesc,func){

    var codes=await codesc 
    var digits=4
    var guessedcodes=codes[0];
    var secretcode=codes[1];
    var HdAndSvdAndcode=[];
    for(var i=0;i<guessedcodes.length;i++)
    {
        //if guessedcode is an array then it's invalid since it's composed from the guessedcode and null value
        if(Array.isArray(guessedcodes[i]))
         {
            HdAndSvdAndcode.push([null,null,guessedcodes[i][0]])
         }
        else
        {
             sumsecret = secretcode
             .split('')
             .map(Number)
             .reduce(function (a, b) {
                 return a + b;
              }, 0);

              nospacecode=guessedcodes[i].trim();
             //calculate the sum of digits of the guessedcode
             sumguessed = nospacecode
             .split('')
             .map(Number)
             .reduce(function (a, b) {
                   return a + b;
                 }, 0);

         var count=0;
    
         for (var j=0;j<digits;j++)
                {
                 if (secretcode.charAt(j)!=nospacecode.charAt(j))
                    {
                     count++;
                    }
                }

         var SVD=sumsecret-sumguessed;
         var HD=count;
         HdAndSvdAndcode.push([HD,SVD,guessedcodes[i]])
        }
    }

    HdAndSvdAndcode.push(secretcode);
     //here at the end of this function we call the next function
     // print
    func(HdAndSvdAndcode,no_op); 
   
}

var print=async function(hdandsvdcodes,func){

        const digits = 4
        const USER_MESSAGE_GUESS_CODE = ">> Guess the %d-digit secret code (%d):";
        const USER_MESSAGE_INVALID_INPUT =   ">> Yikes! Invalid input";
        const USER_MESSAGE_WRONG_GUESS =     ">> Wrong! SVD=%d, HD=%d";
        const USER_MESSAGE_CORRECT_GUESS =   ">> Correct - You won!";
        const USER_MESSAGE_GAME_OVER =       ">> Wrong - Game over!";
        const USER_MESSAGE_CODE =     ">> The secret code was:";

        var codes=await hdandsvdcodes;
        var attempts=10;
        var secretcode=codes[codes.length-1]
        const util = require('util');
        for(var i=0;i<codes.length-1;i++)
        {
             var Hd=codes[i][0];
             var Svd=codes[i][1];
             var guessedcode=codes[i][2];

             console.log(util.format(USER_MESSAGE_GUESS_CODE,digits,attempts));
             console.log(guessedcode+"\n")
             if(attempts==1 && Hd!=null &&Hd!=0)
                {
                    console.log(USER_MESSAGE_GAME_OVER)
                    console.log(util.format(USER_MESSAGE_CODE,secretcode))
                    break;
                }
             else{

                if(Hd==null)
                    console.log(USER_MESSAGE_INVALID_INPUT)
                else if (Hd==0)
                {
                    console.log(USER_MESSAGE_CORRECT_GUESS);
                    break;
                }
                else
                {
                    console.log(util.format(USER_MESSAGE_WRONG_GUESS,Svd,Hd))
                    attempts--;
                }
            }

        }

    //here at the end of this function we call the last function that does nothing
    func(undefined)
}
// ALESSIO why not var here?
no_op=function(func){
    process.exit()
}

extractsecretcode("secret.code",inputGuessedCode);

File: ps-2019-assignment-03-Bessem11-master/task-4/MasterMild.java
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Queue;
import java.util.Scanner;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

public class MasterMild implements Runnable {
    
        //game variables
        private static String secretcode;
        private int digits=4;
    
        //List that containts threads
        private static List<Thread>  workers =new ArrayList<Thread>();
    
        //Two locks to lock shared data so that we don't get incompatoble result
        //when accessing to the shared data
        private static Lock inputLock = new ReentrantLock();
        private static Lock outputLock = new ReentrantLock();
    
    
        //Two data spaces
        private static List<List<String>> Inputs=new ArrayList<List<String>>();
        private static List<List<String>> Outputs=new ArrayList<List<String>>();
    
    
        //games messages
        public static final String USER_MESSAGE_GUESS_CODE =    ">> Guess the %d-digit secret code (%d):\n";
        public static final String USER_MESSAGE_INVALID_INPUT = ">> Yikes! Invalid input";
        public static final String USER_MESSAGE_WRONG_GUESS =   ">> Wrong! SVD=%d, HD=%d\n";
        public static final String USER_MESSAGE_CORRECT_GUESS = ">> Correct - You won!";
        public static final String USER_MESSAGE_GAME_OVER =     ">> Wrong - Game over!";
        public static final String USER_MESSAGE_CODE =     ">> The secret code was: ";
    
    
        //this method merge the partial data and print the final output
    public void print() {
        
                int Attempts=10;
                List<String> codes=new ArrayList<String>();
                int i=0;
        
                while(!Outputs.isEmpty())
                    {
                
                            for(int m=0;m<Outputs.size();m++)
                                {
                                        if(Integer.parseInt(Outputs.get(m).get(2))==i)
                                            {
                                                    //codes[0] containts the HD value, codes[1] contain the SVD value
                                                    //codes[2] contain the rank of the input regarding all the other inputs,codes[3] contain the guessedcode
                                                    codes=Outputs.remove(m);
                                                    i++;
                                                    break;
                                                }
                                    }
                
                            System.out.format(USER_MESSAGE_GUESS_CODE,this.digits,Attempts);
                             System.out.println(codes.get(3)+"\n");
                
                             //for the last attemps if the user doesn't guess the right code
                             //then we didn't print the SVD AND HD
                            if(Attempts==1 &&codes.get(0)!=null&&Integer.parseInt(codes.get(0))!=0)
                                {
                                        System.out.println(USER_MESSAGE_GAME_OVER);
                                        System.out.format(USER_MESSAGE_CODE+secretcode+"\n");
                                        break;
                                    }
                            else
                                {
                        
                                      //If the guessedcode was invalid
                                     if(codes.get(0)==null)
                                        {
                                                System.out.println(USER_MESSAGE_INVALID_INPUT);
                                            }
                                     // If HD=0
                                    else if(Integer.parseInt(codes.get(0))==0)
                                        {
                                                System.out.println(USER_MESSAGE_CORRECT_GUESS);
                                                break;
                                
                                            }
                                    // If the guessedcode was valid but Hd!=0
                                    else
                                        {
                                                System.out.format(USER_MESSAGE_WRONG_GUESS,Integer.parseInt(codes.get(1)),Integer.parseInt(codes.get(0)));
                                                Attempts--;
                                
                                            }
                        
                                    }
                
                
                        }
            }
    
    
        public void extract_secretcode(String SECRET_CODE_LOCATION)  {
        
        
                //Read secret code file
                try
                {
                    InputStream flux=new FileInputStream(SECRET_CODE_LOCATION);
                    InputStreamReader lecture=new InputStreamReader(flux);
                    BufferedReader buff=new BufferedReader(lecture);
                    String ligne;
                    ligne=buff.readLine();
                    secretcode=ligne;
                    }catch(IOException e)
                {
                        System.out.println("No valid path or file not found");
                    }
           }
    
      public void adjustcode() {
        
        
                  //Add zeros
                  if(secretcode.length()<this.digits)
                      {
                              int lengthdiff=this.digits-secretcode.length();
                
                              for (int i=0; i<lengthdiff;i++)
                                  {
                                          secretcode='0'+secretcode;
                                      }
                          }
                  // pick less digits of the secret_code number because the secret_code length is higher then Digits
                  else if ((secretcode.length()>this.digits))
            
                      {
                              int lengthdiff=secretcode.length()-this.digits;
                              secretcode=secretcode.substring(lengthdiff, secretcode.length());
                          }
        
              }
    
    
    
        @Override
        public void run() {
        
                while(true)
                    {
                
                            int hd = 0;
                            int svd=0;
                            String position;
                            boolean isnumber=true;
                            String guessedcode="";
                            List<String> readinput=new ArrayList<String>();
                            List<String> codes=new ArrayList<String>();
                
                            inputLock.lock();
                            if(!Inputs.isEmpty())
                                {
                                        readinput=Inputs.remove(0);
                                        position=readinput.get(0);
                                        guessedcode  = readinput.get(1);
                                        inputLock.unlock();
                                    }
                            else
                                {
                                        inputLock.unlock();
                                        break;
                                    }
                
                
                
                
                
                
                        String noSpaceCode=guessedcode.trim();
                        //verify if the guessed_code is a number
                        for (int i=0 ;i<noSpaceCode.length()&&isnumber;i++)
                                 {
                                        if (!Character.isDigit(noSpaceCode.charAt(i)))
                                                isnumber=false;
                                     }      
                        
                        if (noSpaceCode.length()!=this.digits || !isnumber)
                              {
                                    //here w insert invalide hd and svd to indicate that the Guessecode
                                    //is not valid since we cannot have Hd=0 and SVD=0 they must be both
                                    //equal to zero
                                    codes.add(null);
                                    codes.add(null);
                                    codes.add(position);
                                    codes.add(guessedcode);
                        
                                  }
                        else
                            {
                                int sumguessedcode = (noSpaceCode).chars().map(Character::getNumericValue).sum();
                                // compute the sum of digits of secretcode numbers
                                int sumsecretcode = (secretcode).chars().map(Character::getNumericValue).sum();
                                // compute the SVD value
                                svd = sumsecretcode - sumguessedcode;
                                // compute the HD value
                                for (int i = 0; i < this.digits; i++) {
                                        if ((noSpaceCode).charAt(i) != (secretcode).charAt(i))
                                                hd++;
                                    }
                                
                                codes.add(Integer.toString(hd));
                                codes.add(Integer.toString(svd));
                                codes.add(position);
                                codes.add(guessedcode);
                        
                                }
                        
                        outputLock.lock();
                        try
                          { 
                            Outputs.add(codes);
                              }
                        finally
                        {
                                outputLock.unlock();
                    
                            }
                
                        }
                
                
            }
        
        public static void main(String[] args) throws IOException, InterruptedException {
                
                MasterMild mastermild=new MasterMild();
                //Scanner scanner = new Scanner(System.in);
                BufferedReader in = new BufferedReader(new InputStreamReader(System.in)); 
                mastermild.extract_secretcode("secret.code");
                mastermild.adjustcode();
                String line;
                //Read the guessedcode
                int i=0;
                while ((line = in.readLine()) != null) {
                        List<String> guessedcodes = new ArrayList<String>();
                        guessedcodes.add(Integer.toString(i));
                        guessedcodes.add(line);
                        Inputs.add(guessedcodes);
                        i++;  
                     }
                  
                for(int j=0;j<3;j++) {
                        Thread t1=new Thread(mastermild);
                         workers.add(t1);
                     }
                //Here we  launch them at their jobs
                 for(Thread worker:workers) {
                         worker.start();
                     }
                 for(Thread worker:workers) {
                         worker.join();
                     }
                 
                 
                 //Let's merge the partial results and print the output  
                 mastermild.print();
                 
            }
    
    
    
        
    
    }